File: saps_score.c
/* file: saps_score.c		G. Moody	22 March 2012

Calculate SAPS-I score given data from a Challenge 2012 record
*/

#define PLEN 16	/* max length of data type string, including trailing zero */
struct data {
    int time;   	/* minutes since admission */
    char p[PLEN];		/* type of data */
    double val;		/* parameter value */
};

typedef struct data Data; 

#define MISSING (-2)
#define INVALID (-1)


int score_age(double age)
{
    if     (age > 200) return INVALID;
    else if (age > 75) return 4;
    else if (age > 65) return 3;
    else if (age > 55) return 2;
    else if (age > 45) return 1;
    else if (age >= 0) return 0;
    else               return MISSING;
}

int score_hr(double hr)
{
    if      (hr >  250) return INVALID;
    else if (hr >= 180) return 4;
    else if (hr >= 140) return 3;
    else if (hr >= 110) return 2;
    else if (hr >=  70) return 0;
    else if (hr >=  55) return 2;
    else if (hr >=  40) return 3;
    else if (hr >=  10) return 4;
    else if (hr >=   0) return INVALID;
    else                return MISSING;
}

int score_sbp(double sbp)
{
    if      (sbp >  300) return INVALID;
    else if (sbp >= 190) return 4;
    else if (sbp >= 150) return 2;
    else if (sbp >=  80) return 0;
    else if (sbp >=  55) return 2;
    else if (sbp >=  20) return 4;
    else if (sbp >=   0) return INVALID;
    else                 return MISSING;
}

int score_temp(double temp)
{
    if      (temp >  45) return INVALID;
    else if (temp >= 41) return 4;
    else if (temp >= 39) return 3;
    else if (temp >= 38.5) return 2;
    else if (temp >= 36) return 0;
    else if (temp >= 34) return 1;
    else if (temp >= 32) return 2;
    else if (temp >= 30) return 3;
    else if (temp >= 15) return 4;
    else if (temp >=  0) return INVALID;
    else		 return MISSING;
}

int score_resp(double resp)
{
    if      (resp >  80) return INVALID;
    else if (resp >= 50) return 4;
    else if (resp >= 35) return 3;
    else if (resp >= 25) return 1;
    else if (resp >= 12) return 0;
    else if (resp >= 10) return 1;
    else if (resp >=  6) return 2;
    else if (resp >=  2) return 4;
    else if (resp >=  0) return INVALID;
    else	 	 return MISSING;
}

int score_ur(double ur)
{
    if      (ur > 20.0) return INVALID;
    else if (ur >= 5.0) return 2;
    else if (ur >= 3.5) return 1;
    else if (ur >= 0.7) return 0;
    else if (ur >= 0.5) return 2;
    else if (ur >= 0.2) return 3;
    else if (ur >=   0) return 4;
    else	        return MISSING;
}

int score_bun(double bun)
{
    if      (bun >  100) return INVALID;
    else if (bun >=  55) return 4;
    else if (bun >=  36) return 3;
    else if (bun >=  29) return 2;
    else if (bun >= 7.5) return 1;
    else if (bun >= 3.5) return 0;
    else if (bun >=   1) return 1;
    else if (bun >=   0) return INVALID;
    else	 	 return MISSING;
}

int score_hct(double hct)
{
    if      (hct >  80) return INVALID;
    else if (hct >= 60) return 4;
    else if (hct >= 50) return 2;
    else if (hct >= 46) return 1;
    else if (hct >= 30) return 0;
    else if (hct >= 20) return 2;
    else if (hct >=  5) return 4;
    else if (hct >=  0) return INVALID;
    else	 	return MISSING;
}

int score_wbc(double wbc)
{
    if      (wbc > 200) return INVALID;
    else if (wbc >= 40) return 4;
    else if (wbc >= 20) return 2;
    else if (wbc >= 15) return 1;
    else if (wbc >=  3) return 0;
    else if (wbc >=  1) return 2;
    else if (wbc >=0.1) return 4;
    else if (wbc >=  0) return INVALID;
    else	        return MISSING;
}

int score_glu(double glu)
{
    if      (glu >  1000) return INVALID;
    else if (glu >= 44.5) return 4;
    else if (glu >= 27.8) return 3;
    else if (glu >= 14.0) return 1;
    else if (glu >=  3.9) return 0;
    else if (glu >=  2.8) return 2;
    else if (glu >=  1.6) return 3;
    else if (glu >=  0.5) return 4;
    else if (glu >=    0) return INVALID;
    else	  	  return MISSING;
}

int score_k(double k)
{
    if      (k >   20) return INVALID;
    else if (k >= 7.0) return 4;
    else if (k >= 6.0) return 3;
    else if (k >= 5.5) return 2;
    else if (k >= 3.5) return 0;
    else if (k >= 3.0) return 1;
    else if (k >= 2.5) return 2;
    else if (k >= 0.5) return 4;
    else if (k >=   0) return INVALID;
    else	       return MISSING;
}

int score_na(double na)
{
    if      (na >  200) return INVALID;
    else if (na >= 180) return 4;
    else if (na >= 161) return 3;
    else if (na >= 156) return 2;
    else if (na >= 151) return 1;
    else if (na >= 130) return 0;
    else if (na >= 120) return 2;
    else if (na >= 110) return 3;
    else if (na >=  50) return 4;
    else if (na >=  0)  return INVALID;
    else	 	return MISSING;
}

int score_hco3(double hco3)
{
    if      (hco3 > 100) return INVALID;
    else if (hco3 >= 40) return 4;
    else if (hco3 >= 30) return 1;
    else if (hco3 >= 20) return 0;
    else if (hco3 >= 10) return 1;
    else if (hco3 >=  5) return 2;
    else if (hco3 >=  2) return 4;
    else if (hco3 >=  0) return INVALID;
    else	 	 return MISSING;
}

int score_gcs(double gcs)
{
    if      (gcs >  15) return INVALID;
    else if (gcs >= 13) return 0;
    else if (gcs >= 10) return 1;
    else if (gcs >=  7) return 2;
    else if (gcs >=  4) return 3;
    else if (gcs >=  3) return 4;
    else if (gcs >=  0) return INVALID;
    else	 	return MISSING;
}

int saps_score(Data *d, int n)
{
    int s_age, s_hr, s_sbp, s_temp, s_resp, s_ur, s_bun, s_hct,
	s_wbc, s_glu, s_k, s_na, s_hco3, s_gcs;
    int i, s, saps, missing = 0;
    double urine_tot = 0.;
    
    s_age = s_hr = s_sbp = s_temp = s_resp = s_ur = s_bun = s_hct =
	    s_wbc = s_glu = s_k = s_na = s_hco3 = s_gcs = MISSING;

     /* examine data from the first 24 hours only */
    for (i = 0; i < n && d->time <= 24*60; i++, d++) {
	if (strcmp(d->p, "Age") == 0) {
	    if ((s = score_age(d->val)) > s_age)       s_age = s; }
	else if (strcmp(d->p, "HR") == 0) {
	    if ((s = score_hr(d->val)) > s_hr)         s_hr = s; }
	else if (strcmp(d->p, "SysABP") == 0 || strcmp(d->p, "NISysABP") == 0) {
	    if ((s = score_sbp(d->val)) > s_sbp)       s_sbp = s; }
	else if (strcmp(d->p, "Temp") == 0) {
	    if ((s = score_temp(d->val)) > s_temp)     s_temp = s; }
	else if (strcmp(d->p, "RespRate") == 0) {
	    if ((s = score_resp(d->val)) > s_resp)     s_resp = s; }
	else if (strcmp(d->p, "MechVent") == 0) {
	    if ((s = score_resp(49.0)) > s_resp)       s_resp = s; }
	else if (strcmp(d->p, "Urine") == 0)
	    urine_tot += d->val;
	else if (strcmp(d->p, "BUN") == 0) {
	    if ((s = score_bun(d->val/2.8)) > s_bun)   s_bun = s; }
	else if (strcmp(d->p, "HCT") == 0) {
	    if ((s = score_hct(d->val)) > s_hct)       s_hct = s; }
	else if (strcmp(d->p, "WBC") == 0) {
	    if ((s = score_wbc(d->val)) > s_wbc)       s_wbc = s; }
	else if (strcmp(d->p, "Glucose") == 0) {
	    if ((s = score_glu(d->val/18.0)) > s_glu)  s_glu = s; }
	else if (strcmp(d->p, "K") == 0) {
	    if ((s = score_k(d->val)) > s_k)  	       s_k = s; }
	else if (strcmp(d->p, "Na") == 0) {
	    if ((s = score_na(d->val)) > s_na)         s_na = s; }
	else if (strcmp(d->p, "HCO3") == 0) {
	    if ((s = score_hco3(d->val)) > s_hco3)     s_hco3 = s; }
	else if (strcmp(d->p, "GCS") == 0) {
	    if ((s = score_gcs(d->val)) > s_gcs)       s_gcs = s; }
    }

    /* s_ur is determined by total urine output over 24 hours */
    s_ur = score_ur(urine_tot/1000.0);

    /* Sum the SAPS components. */
    if (s_age   >= 0) saps  = s_age;  else missing++;
    if (s_hr    >= 0) saps += s_hr;   else missing++;
    if (s_sbp   >= 0) saps += s_sbp;  else missing++;
    if (s_temp  >= 0) saps += s_temp; else missing++;
    if (s_resp  >= 0) saps += s_resp; else missing++;
    if (s_ur    >= 0) saps += s_ur;   else missing++;
    if (s_bun   >= 0) saps += s_bun;  else missing++;
    if (s_hct   >= 0) saps += s_hct;  else missing++;
    if (s_wbc   >= 0) saps += s_wbc;  else missing++;
    if (s_glu   >= 0) saps += s_glu;  else missing++;
    if (s_k     >= 0) saps += s_k;    else missing++;
    if (s_na    >= 0) saps += s_na;   else missing++;
    if (s_hco3  >= 0) saps += s_hco3; else missing++;
    if (s_gcs   >= 0) saps += s_gcs;  else missing++;

    if (missing == 0) return (saps);
    else              return (-saps);	/* return partial scores as negative */
}	

File: saps_score.m
function [SAPS_SCORE]=saps_score(varargin)
% [SAPS_SCORE]=saps_score(tm,category,val,truncated)
%
% Calculates SAPS scores. Variables are:
%
%tm      - (Nx1 Cell Array) Cell array containing time of measurement
%category- (Nx1 Cell Array) Cell array containing type (category)
%           measurement
%value   - (Nx1 Cell Array) Cell array containing value of measurement
%truncated - (Logical) Optional flag, if true, will attempt to calculate
%             "truncated" SAPS even if some of the input variables are
%             missing. Default is 0 (false) .
% SAP_SCORE - (Scalar) Value between 0 and 56 representing the severity of
%             the patient's status (higher scores are worse).  A NaN
%             value is returned along with a warning message if the SAPS
%             score cannot be calculated.
%
% Written by Ikaro Silva, 2012
%
% Version 1.0
%
% Reference:
% Gall et al, "A simplified acute physiology score for ICU patients",
% Critical Care Medicine (1984), 12(11).

truncated=0;
inputs={'tm','category','val','truncated'};
for n=1:nargin
    if(~isempty(varargin{n}))
        eval([inputs{n} '=varargin{n};'])
    end
end

%SAPS variable names
SAPS={{'Age'},{'HR'},{'SysABP','NISysABP'},{'Temp'},...
    {'RespRate','MechVent'},{'Urine'},{'BUN'},{'HCT'},...
    {'WBC'},{'Glucose'},{'K'},{'Na'},{'HCO3'},{'GCS'}};

MX_SAPS=56; %Max SAPS value

%Convert SAPS info into tables [min range, max range, SAPS score;...]
%Also convert the units in the table to match the units of the data
Age=[0,45,0;46,55,1;56,65,2;66,75,3;76,200,4];
HR=[70,109,0;55,69,2;110,139,2;40,54,3;140,179,3;180,250,4;10,39,4];
SysABP=[80,149,0;55,79,2;150,189,2;190,300,4;20,54,4];
Temp=[36,38.4,0;34,35.9,1;38.5,38.9,1;32,33.9,2;30,31.9,3;39,40.9,3;41,45,4;15,29.9,4];
RespRate=[12,24,0;10,11,1;25,34,1;6,9,2;35,49,3;2,5,4;50,80,4];
MechVent=49; %Equivalent to a RespRate that will yield a SAPS value of 3
Urine=[5,20,0.002;3.5,4.999,0.001;0.7,3.499,0;0.5,0.699,0.002;0.2,0.499,0.003;0,0.199,0.004].*1000; %Convert from L to mL
BUN=[55,100;36,54.9;29,35.9;7.5,28.9;3.5,7.4;1,3.4].*2.8; %Convert to mg/dL
BUN(:,3)=[4;3;2;1;0;1];
BUN(:,1)=[BUN(2:end,2)+eps;BUN(end,1)];
HCT=[60,80,4;50,59.9,2;46,49.9,1;30,45.9,0;20,29.9,2;5,19.9,4];
WBC=[40,200,4;20,39.9,2;15,19.9,1;3,14.9,0;1,2.9,2;0.100,0.9,4];
Glucose=[44.5,1000;27.8,44.4;14,27.7;3.9,13.9;2.8,3.8;1.6,2.7;0.5,1.5].*18;%Convert to mg/dL
Glucose(:,3)=[4;3;1;0;2;3;4];
Glucose(:,1)=[Glucose(2:end,2)+eps;Glucose(end,1)];
K=[7,20,4;6,6.9,3;5.5,5.9,2;3.5,5.4,0;3,3.4,1;2.5,2.9,2;0.5,2.4,4];
Na=[180,200,4;161,179,3;156,160,2;151,155,1;130,150,0;120,129,2;110,119,3;50,109,4];
HCO3=[40,100,4;30,39.9,1;20,29.9,0;10,19.9,1;5,9.9,3;2,4.9,4];
GCS=[13,15,0;10,12,1;7,9,2;4,6,3;3,3,4];

%Only use data for the first 24 hrs (standard SAPS)
tm=cell2mat(tm);
fr_data=find(str2num(tm(:,1:2))<24);
val=val(fr_data);
tm=tm(fr_data);
category=category(fr_data);

%Loop through all SAPS variables, adding risk points to SAPS_SCORE according to their tables
SAPS_SCORE=NaN;
for s=1:length(SAPS)
    %Get data for the selected category only (If more than one name exist for the variables, merge data)
    saps_var=SAPS{s};
    sig_ind= val.*0;
    eval(['table=' regexprep(saps_var{1},' ','_') ';'])
    for i=1:length(saps_var)
        sig_ind=sig_ind | strcmp(saps_var(i),category);
    end
    tmp_data=val(sig_ind);
    
    if(strcmp(saps_var{1},'RespRate'))
        %For Respiration, check Mechanical Ventilation flag
        %If subject is on ventillation, this is equivalent to having a
        %RespRate that will yield a SAPS value of 3
        tmp_category=category(sig_ind);
        mech_vent_ind=find(strcmp(saps_var(2),tmp_category)==1);
        if(~isempty(mech_vent_ind) && any(mech_vent_ind))
            tmp_data=MechVent;
        end
    end
    if(strcmp(saps_var{1},'Urine'))
        %For Urine output, get cumulative over 24hrs
        tmp_data=sum(tmp_data);
    end
    
    %Set out of bound values to NaN
    tmp_data(tmp_data < min(table(:,1)))=NaN;
    tmp_data(tmp_data > max(table(:,2)))=NaN;
    %Get the lowest and highest values for the measured variable
    if(all(isnan(tmp_data)) || isempty(tmp_data))
        if(~truncated)
            warning(['No SAPS score for variable ' saps_var{:} '... exiting.'])
            SAPS_SCORE=NaN;
            break
        else
            warning(['No variable ' saps_var{:} '... attempting truncated SAPS.'])
        end
    else
        mn=nanmin(tmp_data);
        mn_saps_ind=find( ( mn(1) >= table(:,1) & mn(1) <= table(:,2) ) ==1);
        mn_saps_val=table(mn_saps_ind,3);
        mx=nanmax(tmp_data);
        mx_saps_ind=find( ( mx(1) >= table(:,1) & mx(1) <= table(:,2) ) ==1);
        mx_saps_val=table(mx_saps_ind,3);
        %Add worst case scenario to SAPS SCORE
        if(isnan(SAPS_SCORE))
            SAPS_SCORE= nanmax(mn_saps_val,mx_saps_val);
        else
            SAPS_SCORE= SAPS_SCORE + nanmax(mn_saps_val,mx_saps_val);
        end
    end
end


